executor/aggregate.go

```diff
@@ -862,6 +871,56 @@ func (e *HashAggExec) getPartialResults(groupKey string) []aggfuncs.PartialResul
 	return partialResults
 }
 
+func (e *HashAggExec) executeFirstRow(ctx context.Context, chk *chunk.Chunk) (err error) {
+	for {
+		if e.firstRowProcessed == e.childResult.NumRows() {
+			mSize := e.childResult.MemoryUsage()
+			e.firstRowProcessed = 0
+			err := Next(ctx, e.children[0], e.childResult)
+			e.memTracker.Consume(e.childResult.MemoryUsage() - mSize)
+			if err != nil {
+				return err
+			}
+			// no more data.
+			if e.childResult.NumRows() == 0 {
+				return nil
+			}
+			e.groupKeyBuffer, err = getGroupKey(e.ctx, e.childResult, e.groupKeyBuffer, e.GroupByItems)
+			if err != nil {
+				return err
+			}
+		}
+
+		failpoint.Inject("unparallelHashAggError", func(val failpoint.Value) {
+			if val.(bool) {
+				failpoint.Return(errors.New("HashAggExec.unparallelExec error"))
+			}
+		})
+
+		for e.firstRowProcessed < e.childResult.NumRows() {
+			groupKey := string(e.groupKeyBuffer[e.firstRowProcessed]) // do memory copy here, because e.groupKeyBuffer may be reused.
+			if !e.groupSet.Exist(groupKey) {
+				e.groupSet.Insert(groupKey)
+				partialResults := e.getPartialResults(groupKey)
```

Comment:
getPartialResults will also handle the result in partialResultMap. But in the optimization, it is useless. It will occupy some memory until the executor close.

I think we can allocPartialResult for each aggFuncs, in L906.  
```
	for _, af := range e.PartialAggFuncs {
					partialResult, _ := af.AllocPartialResult()
					_, err := af.UpdatePartialResult(e.ctx, []chunk.Row{e.childResult.GetRow(e.firstRowProcessed)}, partialResult)
					if err != nil {
						return err
					}
					if err := af.AppendFinalResult2Chunk(e.ctx, partialResult, chk); err != nil {
						return err
					}
				}
```

Of course, don't track the memory usage is also ok, I think. Because it the result will use once, and it will be GC by golang later.